plugins/parquet/parquetembed.hpp

```diff
@@ -0,0 +1,1082 @@
+        // Create a json string of all the fields which will be much more performant.
+        r_parquet->begin_row();
+        return true;
+    }
+    virtual void processEndSet(const RtlFieldInfo *field)
+    {
+        r_parquet->end_row(field->name);
+    }
+    virtual void processEndDataset(const RtlFieldInfo *field)
+    {
+        UNSUPPORTED("DATASET");
+    }
+    virtual void processEndRow(const RtlFieldInfo *field)
+    {
+        r_parquet->end_row(field->name);
+    }
+
+protected:
+    inline unsigned checkNextParam(const RtlFieldInfo *field);
+
+    const RtlTypeInfo *typeInfo = nullptr;
+    const IContextLogger &logctx;
+    int firstParam;
+    RtlFieldStrInfo dummyField;
+    int thisParam;
+    TokenSerializer m_tokenSerializer;
+
+    std::shared_ptr<ParquetHelper> r_parquet;
+    bool partition; //! Local copy of a boolean so we can know if we are writing partitioned files or not.
+};
+
+/**
+ * @brief Binds an ECL dataset to a vector of parquet objects.
+ *
+ */
+class ParquetDatasetBinder : public ParquetRecordBinder
+{
+public:
+    /**
+     * @brief Construct a new ParquetDataset Binder object
+     *
+     * @param _logctx logger for building the dataset.
+     * @param _input Stream of input of dataset.
+     * @param _typeInfo Field type info.
+     * @param _query Holds the builder object for creating the documents.
+     * @param _firstParam Index of the first param.
+     */
+    ParquetDatasetBinder(const IContextLogger &_logctx, IRowStream *_input, const RtlTypeInfo *_typeInfo, std::shared_ptr<ParquetHelper> _parquet, int _firstParam)
+        : input(_input), ParquetRecordBinder(_logctx, _typeInfo, _firstParam, _parquet)
+    {
+        d_parquet = _parquet;
+        // getFieldTypes(_typeInfo);
+
+        reportIfFailure(d_parquet->fieldsToSchema(_typeInfo));
+    }
+    virtual ~ParquetDatasetBinder() = default;
+    void getFieldTypes(const RtlTypeInfo *typeInfo);
+
+    /**
+     * @brief Gets the next ECL row.
+     *
+     * @return true If there is a row to process.
+     * @return false If there are no rows left.
+     */
+    bool bindNext()
+    {
+        roxiemem::OwnedConstRoxieRow nextRow = (const byte *)input->ungroupedNextRow();
+        if (!nextRow)
+            return false;
+        processRow((const byte *)nextRow.get()); // Bind the variables for the current row
+        return true;
+    }
+
+    void writeRecordBatch()
+    {
+        // convert row_batch vector to RecordBatch and write to file.
+        PARQUET_ASSIGN_OR_THROW(auto record_batch, d_parquet->ConvertToRecordBatch(*(d_parquet->record_batch()), d_parquet->getSchema()));
+        // Write each batch as a row_groups
+        PARQUET_ASSIGN_OR_THROW(auto table, arrow::Table::FromRecordBatches(d_parquet->getSchema(), {record_batch}));
+
+        if (partition)
+        {
+            reportIfFailure(d_parquet->writePartition(table));
+        }
+        else
+        {
+            CriticalBlock block(fileLock);
```

Comment:
It was included inside of the writePartition function above rather than in the if else block. I have removed the critical section because the workers are not writing to the same file.

Reply:
There shouldn't be any multithreading issues. Multithreaded reading of files is supported in Arrow, and when writing parquet files each worker has its own file.